dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java

```diff
@@ -1412,16 +1526,22 @@ private boolean pruneData(ThroughputDistribution priorStats, double forecast, in
      * @return - smallest valid poolSize found
      */
     private Integer getSmallestValidPoolSize(Integer poolSize, Double forecast, int compareSpan) {
-        Integer smallestPoolSize = -1;
+        Integer smallestPoolSize = threadStats.firstKey();
+        Integer nextPoolSize = threadStats.higherKey(smallestPoolSize);
+        Integer pruneSize = -1;
         boolean validSmallData = false;
         int distance = 0;
         while (!validSmallData) {
-            smallestPoolSize = threadStats.firstKey();
             ThroughputDistribution smallestPoolSizeStats = getThroughputDistribution(smallestPoolSize, false);;
             distance = poolSize - smallestPoolSize;
             // prune data that is too old or outside believable range
             if (pruneData(smallestPoolSizeStats, forecast, distance, compareSpan)) {
-                threadStats.remove(smallestPoolSize);
+                pruneSize = smallestPoolSize;
+                smallestPoolSize = nextPoolSize;
+                nextPoolSize = threadStats.higherKey(smallestPoolSize);
+                if (pruneSize > hangBufferPoolSize) {
+                    threadStats.remove(pruneSize);
+                }
```

Comment:
Ummm, no ... this is sort of subtle, I hope I got the logic right ... 
Background: The basic concept of getSmallestValidPoolSize is to find a trustworthy comparison datapoint for a poolsize smaller than the current one, as a setup to doing a linear approximation to see if adding more threads is likely to be helpful. The prior implementation would discard all smaller poolsize data that was deemed untrustworthy by pruneData. 
However ... in the case of a threadpool hang scenario, smaller poolsize data that looks untrustworthy or irrelevant (e.g. orders of magnitude lower tput) may in fact be telling the truth - if we drop the pool to that size, we really will see tiny (maybe no) throughput because we will fall below the hang threshold.
So ... the change here is to NOT prune that down-side (smaller poolsize) data if it is below a previously observed hang threshold. That way we keep the smaller poolsize data in the regular evaluation process, and if it has really low tput values then it will discourage the regular controller process from shrinking to those poolsizes.
However - although we are keeping that data around, it is not 'valid' from the standpoint of the linear approximation that we want to make, so we do NOT set validSmallData true ... As we keep moving up the poolsizes, we may find valid data in some intermediate smaller poolsize which is far enough away from the current poolsize to perform the linear approximation, or we may not find any valid data till we get to the current poolsize, which will always return false from pruneData and thus cause validSmallData to be set true.

Please let me know if this doesn't make sense to you, it might mean I haven't explained it well enough or maybe the logic itself is incorrect somehow ... ;-)